[SR-755] Tool for re-symbolicating fatal stacktraces on Linux.

gmilos opened the pull request:
<!-- Please complete this template before creating the pull request. -->
#### What's in this pull request?
On Linux stack traces are full of <unavailable> symbols. This provides a tool to re-symbolicate such stack trace. 

<!-- Description about pull request. -->
The cause is difference in behaviour of `dladdr` between Linux & Darwin. See SR.

#### Resolved bug number: ([SR-755](https://bugs.swift.org/browse/SR-755))
<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [x] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        All supported platforms     | @swift-ci Please smoke test and merge
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

A smoke test on macOS does the following:

1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:

1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        All supported platforms     | @swift-ci Please test and merge
        OS X platform               | @swift-ci Please test OS X platform
        OS X platform               | @swift-ci Please benchmark
        Linux platform              | @swift-ci Please test Linux platform


**Lint Testing**

        Language     | Comment
        ------------ | -------------
        Python       | @swift-ci Please Python lint

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

DougGregor commented:
@swift-ci please smoke test

gmilos commented:
The failure on Linux was a timeout, doesn't seem to be related to the change I made. Trying to kick off another round of ci.

gmilos commented:
@swift-ci please smoke test

gmilos commented:
@swift-ci Please test Linux platform

DougGregor commented:
@swift-ci please smoke test Linux

modocache commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -72,7 +72,6 @@ static bool getSymbolNameAddr(llvm::StringRef libraryName, Dl_info dlinfo,
   // If the address is unavailable, just use <unavailable> as the symbol
```
nit-pick: Since you remove the "<unavailable>" string below, this comment here is no longer accurate. Could you update this comment as well?

gribozavr commented on utils/symbolicate-linux-fatal.py:
```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env python
```
Please rename the tool omitting the `.py` suffix.  Users have no necessity to know which language the tool is written in.

modocache replied:
This file needs a license header. Take a look at the other Python files in the repository for examples.

gribozavr commented on utils/symbolicate-linux-fatal.py:
```diff
@@ -0,0 +1,119 @@
+            print "{0:s}    {1:s} {2:s}".format(frame_fragment, symbol_fragment, line_fragment)
+        frame_idx = frame_idx + 1
+
+if len(sys.argv) != 3:
```
Please add a `main()` function (by convention used in all other scripts).

gribozavr commented on utils/symbolicate-linux-fatal.py:
```diff
@@ -0,0 +1,119 @@
+exit_code = lddprocess.wait()
+if exit_code != 0:
+    print("ldd failed with {0:d}".format(exit_code))
+    sys.exit(1)
```
`subprocess.check_output`?

gribozavr commented on utils/symbolicate-linux-fatal.py:
```diff
@@ -0,0 +1,119 @@
+# * search symbols by name for the not <unavailable> ones
+
+import sys
+import lldb
```
This LLDB is probably the system one, not the Swift-enabled LLDB (so using it is as good as any other tool).  Did you consider using `addr2line`, which is the usual way of symbolicating?

gm281 replied:
It'll pick up whatever is available in `PYTHONPATH`. If that's the standard (=non-Swift enabled) you're right we'll get as much as we get from `addr2line`. However, if the Swift-enabled one is available, we'll get symbols demangled etc. I'd hope longer term mainstream `lldb` will get Swift support. But `addr2line` probably won't.

Also, `lldb` approach _could_ be used on Darwin too (if there is some need for it).

What do you think @gribozavr ?

gmilos replied:
Oops, was signed in with wrong account (just above).

Last note: I'm more than happy to rewrite with `addr2line` but seems like less promising approach.

gribozavr replied:
> If that's the standard (=non-Swift enabled) you're right we'll get as much as we get from addr2line.

Fair enough.  Seems like the right long-term trade-off.

modocache commented:
You'll want to run the Python linter on this new code, which contains a few violations. There's a handy utility to run the linter, just run `utils/python-lint`.

modocache commented on utils/symbolicate-linux-fatal.py:
```diff
@@ -0,0 +1,119 @@
+
+dynlibs = {}
+def process_ldd(lddoutput):
+    global dynlibs
```
This line, `global dynlibs` is not needed. Since you define `dynlibs` in a global scope above, you are able to reference it throughout this file. This is why you're able to reference it in the `process_stack` function below.

Also, I believe the Python linter will raise this issue, but I think that typically Python globals tend to be in `UPPER_SNAKE_CASE`.

modocache commented on utils/symbolicate-linux-fatal.py:
```diff
@@ -0,0 +1,119 @@
+    sys.exit(1)
+
+binary = sys.argv[1]
+logfile = sys.argv[2]
```
I think it might be better to use the `argparse` module here. That will allow you to specify positional arguments to this script, along with help text that will be displayed if someone invokes the script with the wrong arguments.

modocache commented on utils/symbolicate-linux-fatal.py:
```diff
@@ -0,0 +1,119 @@
+            dynlib_baseaddr = framePC - symbol_offset
+            if dynlib_path in memmap:
+                if memmap[dynlib_path] != dynlib_baseaddr:
+                    print "Mismatched base address for: {0:s}, had: {1:x}, now got {2:x}".format(dynlib_path, memmap[dynlib_path], dynlib_baseaddr)
```
To make this script compatible with Python 3, you should `from __future__ import print_function` above, then use `print()` here.

gmilos commented:
@modocache re utils/python-lint. I run it + fixed up all the issues it found. Thanks for the suggestion.

gmilos commented:
@DougGregor could you kick off another round of tests? I've addressed all the comments, so hopefully this is ready to be merged, pending successful tests.

modocache commented on utils/symbolicate-linux-fatal:
```diff
@@ -0,0 +1,158 @@
+
+
+def main():
+    global BINARY
```
Instead of storing `args.binary` below into this global `BINARY` variable, I would suggest passing `binary` as an argument into the `create_lldb_target` and `process_stack` functions. I think doing so limits the scope where `BINARY` is used and makes the control flow of the program easier to understand.

modocache commented on utils/symbolicate-linux-fatal:
```diff
@@ -0,0 +1,158 @@
+import lldb
+import subprocess
+
+DYN_LIBS = {}
```
As with `BINARY`, I think this program would be easier to understand if `DYN_LIBS` was instantiated in `main()`, then passed in as an argument to the `process_ldd()` and `process_stack()` functions.

modocache commented on utils/symbolicate-linux-fatal:
```diff
@@ -0,0 +1,158 @@
+        if dynlib_fname in DYN_LIBS:
+            dynlib_path = DYN_LIBS[dynlib_fname]
+        else:
+            if dynlib_fname in BINARY:
```
I think this `else: if:` could be combined into an `elif`.

modocache commented on utils/symbolicate-linux-fatal:
```diff
@@ -0,0 +1,158 @@
+                    print("Mismatched base address for: {0:s}, "
+                          "had: {1:x}, now got {2:x}".format(
+                            dynlib_path, memmap[dynlib_path], dynlib_baseaddr))
+                    sys.exit(1)
```
This isn't necessarily better than your approach, but an alternative to printing and calling `sys.exit()` here would be to throw an exception, such as `throw Error("Mismatched ...")`. It's fewer lines of code, and emits a backtrace, but maybe you don't want users to see that. Up to you!

gmilos commented:
@modocache All comments addressed. @DougGregor ready for CI again.

gribozavr commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -132,12 +129,21 @@ static void dumpStackTraceEntry(unsigned index, void *framePC) {
+  if (foundSymbol) {
+    static const char *backtraceEntryFormat = "%-4u %-34s 0x%0.16lx %s + %td\n";
+    fprintf(stderr, backtraceEntryFormat, index, libraryName.data(), symbolAddr,
+          symbolName.c_str(), ptrdiff_t(uintptr_t(framePC) - symbolAddr));
```
In C++ code, the line after the linebreak should be aligned with the opening paren.  (Here, `symbolName` should start in the same column as `stderr`.)

gribozavr commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -132,12 +129,21 @@ static void dumpStackTraceEntry(unsigned index, void *framePC) {
+                                              "<unavailable> + %td\n";
+    fprintf(stderr, backtraceEntryFormat, index, libraryName.data(),
+        uintptr_t(framePC),
+        ptrdiff_t(uintptr_t(framePC) - uintptr_t(dlinfo.dli_fbase)));
```
Formatting (like above).

gribozavr commented:
The code changes generally LGTM, but it would be good to add some tests.

gribozavr commented:
@swift-ci Please smoke test

gribozavr commented:
@swift-ci Please Python lint

gribozavr commented:
@swift-ci Please smoke test

gribozavr commented:
@swift-ci Please Python lint

rintaro commented:
@gmilos
Could you add `symbolicate-linux-fatal` entry in [`.pep8` file](https://github.com/apple/swift/blob/master/.pep8)?

gmilos commented:
Note python lint failed on things unrelated to this pr:
```
--- Python lint ---
+ cd swift
+ ./utils/python-lint
./utils/resolve-crashes.py:19:80: E501 line too long (94 > 79 characters)
./utils/update-checkout:46:80: E501 line too long (94 > 79 characters)
./utils/update-checkout:48:80: E501 line too long (104 > 79 characters)
./utils/update-checkout:51:24: E128 continuation line under-indented for visual indent
Build step 'Execute shell' marked build as failure
```

rintaro commented:
@swift-ci Please Python lint

rintaro commented:
@swift-ci Please Python lint

gmilos commented:
Added to [.pep8](../blob/master/.pep8) and fixed up whitespace in [Errors.cpp](../blob/master/stdlib/public/runtime/Errors.cpp).

@gribozavr I was looking at testing even before you mentioned it. I need to figure out the whole testing platform (that's on me) but also this test will be slightly more complex than average, as it has dependency on lldb (so I need to find a way of referring to lldb python bindings). I found [backtrace test](../blob/master/test/Runtime/backtrace.swift) which is a good template for my test, except of the lldb dependency again. Do you have an example of how I'd express that?

I'd be keen to get this PR in asap, and write the test in a subsequent PR (as there are ppl waiting for it).

rintaro commented on utils/symbolicate-linux-fatal:
```diff
@@ -0,0 +1,155 @@
+
+import argparse
+import lldb
+import subprocess
```
```
./utils/symbolicate-linux-fatal:28:1: I201 Missing newline before sections or imports.
./utils/symbolicate-linux-fatal:29:1: I100 Import statements are in the wrong order. import subprocess should be before import lldb
./utils/symbolicate-linux-fatal:29:1: I201 Missing newline before sections or imports.
```

`import` block should be:
```python
from __future__ import print_function

import argparse
import subprocess

import lldb
```
because `lldb` is not a standard library.

gmilos commented:
@rintaro I noticed I* too. Seems like flake8-import-order doesn't work on my mac for some reason, verified on another machine.

rintaro commented:
@gmilos Thank you!
@swift-ci Please Python lint

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
Python lint result here
https://ci.swift.org/view/Pull%20Request/job/swift-PR-python-lint/91/console

gmilos commented:
@gribozavr @modocache @rintaro Is there anything holding up the merge here?

modocache commented:
@gribozavr mentioned a few days ago that it'd be nice to have some tests, and I agree! `symbolicate-linux-fatal` seems like it would be a great candidate for a few tests.

You can find examples of Python tests [here](https://github.com/apple/swift/tree/a8c48d25efb1266604eab12617ae8eb595c53bb2/validation-test/Python). I'm a fan of using Python's `unittest` module, like `cmpcodesize.swift` and `swift_build_support.swift` do. Alternatively, you could have your test run `symbolicate-linux-fatal` and verify the output is what you expect.

I think adding tests would benefit everyone. They could automate whatever method you've been using locally to make sure `symbolicate-linux-fatal` works. We'd also be able to verify this script works before merging it.

gmilos commented:
@modocache I'm trying to figure out how to express the dependency on lldb from my test. So far I managed to figure out the basics of LIT, as used in this repo, but I'm not too sure it's ever going to be possible to depend on lldb being built (maybe it could be handled with `REQUIRES`?). Alternatively, the symbolication script could be moved to the lldb repo (I'd much prefer to keep it here though, it'll make it a lot more discoverable).

modocache commented:
Yeah, you could potentially use `REQUIRES:`, I think. I don't see anything resembling `REQUIRES: lldb`, though. The requirements are set up in the `test/lit.cfg` and `validation-test/lit.cfg` files. It might be possible to determine whether lldb is present from within those, then set up the requirement.

If this is prohibitively difficult, I'm fine with putting this off until later, and adding a test in a future commit. @gribozavr may have more thoughts here.

gmilos commented:
@modocache After bit of tinkering I managed to plumb through relevant `REQUIRES` (and lldb build dir info). Please see the latest [commit](https://github.com/apple/swift/pull/4479/commits/9d54338ae59c4b84713d4c12e22aa7275ab53c11).

Would be good to check the test does get to run in CI here (it works ok in my setup, but there my be subtle config differences).

I think this end-to-end integration test is superior to unit tests alone. But I could add some of those too. Anything in particular you'd like to see?

modocache commented on test/lit.site.cfg.in:
```diff
@@ -76,6 +74,12 @@ if "@CMAKE_GENERATOR@" == "Xcode":
+    config.available_features.add('lldb')
+    for root, dirs, files in os.walk("@LLDB_BUILD_DIR@"):
+        if root.endswith("site-packages"):
+            config.substitutions.append(('%lldb-python-path', root))
```
After providing a substitution, shouldn't this `break` out of the loop?

modocache replied:
Also, I'm not sure I understand. Where is the `%lldb-python-path` substitution being used? Is it used anywhere? Did you perhaps forget to include the tests in your commit?

modocache replied:
Also, I *think* @ddunbar prefers substitutions like `%{lldb-python-path}`, but I might be wrong.

gmilos replied:
Ah, indeed, test now attached. Good catch, thanks!

Re: break, will add that. In practice it doesn't matter (the build dir for lldb is shallow and contains exactly one match).

Re: {} in substitutions: {} don't seem to be used in current lit substitutions, see [lit.cfg](https://github.com/apple/swift/blob/master/test/lit.cfg#L337).

modocache replied:
It's used in *some* substitutions there -- note `%{python}` -- but I'll defer to @ddunbar.

gmilos replied:
Ack, will take the cue from @ddunbar. I don't mind either way, but matched it with all other substitutions used in my test.

rintaro replied:
I think, we want to keep `lit.site.cfg.in` as simple as possible.
Could you move this logic to `test/lit.cfg`?
And, we don't need `LLDB_ENABLE` flag. 
i.e:
```python
# lit.site.cfg.in
config.lldb_build_root = "@LLDB_BUILD_DIR@"

# lit.cfg
if config.lldb_build_root != "":
    config.available_features.add('lldb')
    # .. discovery
    config.substitutions.append(('%lldb-python-path', discovered_path))
```

modocache commented on utils/build-script-impl:
```diff
@@ -1240,6 +1240,18 @@ if [[ ! "${SKIP_BUILD_SWIFTPM}" ]] ; then
 fi
 
+# Checks if a given product is enabled (i.e. part of $PRODUCS array)
+function contains_product() {
```
nit-pick: Since this is only used in one spot below, I personally wouldn't split it out into its own function. Your call, though!

modocache replied:
nit-pick: Since this is only used in one spot below, I personally wouldn't split it out into its own function. Your call, though!

gmilos replied:
This is called from a lengthy, complex switch statement, would prefer to keep this (unrelated) logic out of that.

ddunbar replied:
I do have mild preference for %{} but lit doesn't enforce it. I personally use it because I think it makes the behavior more obvious, and I had some idea that one day I might refactor lit to require it so that it could have a more strict shell parsing model. Chances of that actually happening... low. :)

modocache commented on test/Runtime/linux-fatal-backtrace.swift:
```diff
@@ -0,0 +1,23 @@
+
+// REQUIRES: executable_test
+// REQUIRES: OS=linux-gnu
+// REQUIRES: lldb
```
Awesome!! 🎉

modocache commented on utils/backtrace-check:
```diff
@@ -18,44 +18,64 @@
+formatting.""")
+    parser.add_argument(
+        "-u", "--check-unavailable", action='store_true',
+        help="Checks if any symbols were unavailable")
```
This seems reasonable to me! :)

Another approach would be to add `CHECK:` statements in your lit test above. But I think there are other tests in the Swift repository that use this style as well (`gyb`, for example).

modocache commented:
It looks like smoke tests don't build lldb on macOS; I'm going to go ahead and run the entire thing.

@swift-ci please test

modocache commented:
It looks like smoke tests don't build lldb on macOS; I'm going to go ahead and run the entire thing.

@swift-ci Please test

gmilos commented:
Right, I'm still don't fully understand at what point the list of tests gets constructed (and filtered with by relevant `REQUIRES`). It seems like it's at least 2 step process. As per the earlier comment, I want to make sure the test gets to run on default (full) build.

rintaro commented on utils/build-script-impl:
```diff
@@ -2039,6 +2051,15 @@ for host in "${ALL_HOSTS[@]}"; do
                     )
                 fi
 
+                if contains_product "lldb" ; then
```
I think just `if [[ ! "${SKIP_BUILD_LLDB}" ]] ; then` is OK.
(no need to create new `contains_product()` function)

gmilos replied:
I realise that at the moment that's equivalent (`SKIP_BUILD_LLDB` is inverse of `contains_product "lldb"`). However, what I'm trying to express/check is that lldb is enabled. At the moment there is only one way to disable (`SKIP`), but IMHO it's less fragile to explicitly check if it's enabled (it'll handle another mechanism for disabling lldb, e.g. due to failed dependency).

Also, the return value from `contains_product` was intentional. Returning 0 on success would make the above if statement read:
`if ! contains_product "lldb"`
That looks wrong.

What do you think @rintaro ?

rintaro replied:
script
```sh
function foo() {
    return 1
}

if foo "lldb"; then
    echo "SUCCESS"
else
    echo "FAIL"
fi
```
prints `FAIL` in my environment.

gmilos replied:
Ahh, much appreciated @rintaro, my f$@#-up. Could that be the source of my issues ... Retesting (I have to figure out how the hell my local tests worked at all!).

gmilos commented:
The failure is due to the `%lldb-python-path` substitution not having worked. I'll try to figure out if it's due to something not having been rebuilt, or an issue searching for the Swig bindings.

rintaro commented on utils/symbolicate-linux-fatal:
```diff
@@ -0,0 +1,162 @@
+    if args.log == "-":
+        logfile = sys.stdin
+    else:
+        logfile = open(args.log, 'rU')
```
Just a Python `argparse` tip:
```
parser.add_argument(
    "log", type=argparse.FileType("rU"), ...)
```
is the shorthand for this.
see https://docs.python.org/2/library/argparse.html#filetype-objects

gmilos replied:
Neat, applying that change.

gmilos commented:
@modcache The build/test log is difficult to work with as I don't see individual job steps, hence I cannot repro the issue locally and/or cannot piece what commands actually got run. I believe I'd need admin account on Jenkins to see that. Can you?

Any other (trial and error) solution I can see is to sprinkle lot more prints in the (python) build scripts, cmake and lit. Run again. Analyse the output. Will be painful.

rintaro commented on utils/build-script-impl:
```diff
@@ -1240,6 +1240,18 @@ if [[ ! "${SKIP_BUILD_SWIFTPM}" ]] ; then
+      return 1
+    fi
+  done
+  return 0
```
`0` -> `1` and `1` -> `0`. In shell script, `0` is success.

rintaro commented:
Even in (not smoke) test, it seems lldb is not built.
https://ci.swift.org/job/swift-PR-Linux/3733/consoleText
```
/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift/utils/build-script: \
    note: using preset 'buildbot_linux_1404_no_lldb', which expands to 
```

gmilos commented:
@rintaro do you have access to Jenkins job settings? I've no idea where 'buildbot_linux_1404_no_lldb' comes from (and what other options exist).

There is a failure in the way my changes add the ability to set `REQUIRES: lldb`. The idea was that if lldb isn't build, that check will fail (and test won't even be run). This doesn't seem to be happening. I'll try to repro + fix this.

However, even with the above, the work won't be completed. The test will only make sense if `REQUIRES: lldb` succeeds on 'normal' builds+some tests.

rintaro commented:
@gmilos No, I have no access to Jenkins machine.
As for `buildbot_linux_1404_no_lldb`, see
https://github.com/apple/swift/blob/master/utils/build-presets.ini#L669
CI is setup to use this preset,
If you want to invoke that yourself, you can:

```
$ utils/build-script --preset=buildbot_linux_1404_no_lldb
```

I think return value of `contains_product` explains the failure.
`--skip-lldb` is specified, but `LLDB_ENABLE` is `TRUE` and actually the `lldb_build_dir` does not exists.

gmilos commented:
Right, I'm testing right now. My working theory is that I had some un-cleaned build of lldb which made the test succeed (despite the fact the actual test invocation was done without lldb).

gmilos commented:
@modocache @gribozavr could you kick off another smoke test please. With the latest fix it should work. However, likely, it won't run the test (due to preset `buildbot_linux_1404_no_lldb` being used).

modocache commented:
You're sure you want a smoke test, not a validation test? I read their descriptions (part of your PR message, if you expand the fold), and it seems like lldb isn't built for macOS when smoke testing. In any case, here goes!

@swift-ci please smoke test

modocache commented:
@swift-ci please smoke test

gmilos commented:
@modocache I don't have access to job configuration on Jenkins, is there a page somewhere describing what goes into smoke vs validation tests vs other jobs? Otherwise I could try scraping that from existing logs, but that's inefficient.

gmilos commented:
Green ticks ... good start. @modocache @gribozavr could you select some build + test with lldb too? Hopefully there shouldn't be anything blocking merge after that.

modocache commented:
> I don't have access to job configuration on Jenkins, is there a page somewhere describing what goes into smoke vs validation tests vs other jobs?

Yup! It's actually on your original pull request. Click on the dropdown for an explanation:

![screen shot 2016-09-09 at 11 26 38 am](https://cloud.githubusercontent.com/assets/552921/18392597/5a8013ce-7680-11e6-95de-670a05515d6a.png)
![screen shot 2016-09-09 at 11 26 47 am](https://cloud.githubusercontent.com/assets/552921/18392596/5a7cab94-7680-11e6-938a-015407327fb1.png)

> could you select some build + test with lldb too?

@swift-ci please test

modocache commented:
@swift-ci please test

modocache commented:
Looks like the newly added test, `Runtime/linux-fatal-backtrace.swift`, failed.

gmilos commented:
@modocache the build is now gone :(, could you please rebuild again. I'll inspect the failure once the output is available.

modocache commented:
@swift-ci please test

rintaro commented:
@swift-ci Please smoke test

modocache commented:
The tests appear to have passed, but in reality the new test was never run. From the logs:

```
12:04:04 UNSUPPORTED: Swift(linux-x86_64) :: Runtime/linux-fatal-backtrace.swift (266 of 8611)
```

gmilos commented:
@modocache @rintaro There was a bug working out if `lldb` `PRODUCT` is enabled. This is now fixed. Could you please kick off Smoke Tests (where I expect the test to be run) and regular tests (where I don't). I have now tested this locally (I finally understand enough about how the presets work to reproduce, hopefully equivalent, test run locally).

gmilos commented:
@modocache @rintaro Could either one of you kick off Smoke Tests + regular Tests? (I don't have the rights to do that :( ).

gmilos commented:
@DougGregor @gribozavr Any chance either one of you could kick off Smoke + regular tests please?

gribozavr commented:
@swift-ci Please smoke test

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please smoke test

gribozavr commented:
@swift-ci Please test

DougGregor commented:
Tests pass, but now we have a conflict. Argh! @gmilos , can you clear up the conflict so we can merge?

gmilos commented:
@DougGregor yep, simple conflict in `.pep8` now fixed (hit that before too).

Another round of tests would be very useful. Could you kick them off for me please? Alternatively, if you added me to the Apple group, I could make sure the tests are all green before asking for review/merge.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please test

gmilos commented:
Build failure is unrelated to the patch:

```
15:56:37 FAILED: /usr/bin/clang++   -DCLANG_ENABLE_ARCMT -DCLANG_ENABLE_OBJC_REWRITER -DCLANG_ENABLE_STATIC_ANALYZER -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools/clang/lib/CodeGen -I/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/llvm/tools/clang/lib/CodeGen -I/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/llvm/tools/clang/include -Itools/clang/include -Iinclude -I/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/llvm/include -fno-stack-protector -fPIC -fvisibility-inlines-hidden -Wall -W -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wcovered-switch-default -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Werror=date-time -std=c++11 -fcolor-diagnostics -ffunction-sections -fdata-sections -fno-common -Woverloaded-virtual -Wno-nested-anon-types -O3    -UNDEBUG  -fno-exceptions -fno-rtti -MMD -MT tools/clang/lib/CodeGen/CMakeFiles/clangCodeGen.dir/CGDebugInfo.cpp.o -MF tools/clang/lib/CodeGen/CMakeFiles/clangCodeGen.dir/CGDebugInfo.cpp.o.d -o tools/clang/lib/CodeGen/CMakeFiles/clangCodeGen.dir/CGDebugInfo.cpp.o -c /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/llvm/tools/clang/lib/CodeGen/CGDebugInfo.cpp
15:56:37 
/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/llvm/tools/clang/lib/CodeGen/CGDebugInfo.cpp:3572:64: error: cannot initialize a parameter of type 'llvm::DIExpression *' with an lvalue of type 'llvm::GlobalVariable *'
15:56:37                                        Var->hasLocalLinkage(), Var, nullptr);
```

gmilos commented:
Same failure is present here:
https://ci.swift.org/job/oss-swift-incremental-RA-linux-ubuntu-15_10/7724/consoleFull#-1191272582ee1a197b-acac-4b17-83cf-a53b95139a76
(that shows the offending set of changes too).

@DougGregor would you mind kicking of another set of tests on this PR (both smoke and regular) once the above llvm issue is fixed?

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
Linux PR test passed here: https://ci.swift.org/view/Pull%20Request/job/swift-PR-Linux-smoke-test/1340/

macOS PR test passed here: https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx-smoke-test/1845/
